### PR TITLE
[SPARK-48011][Core] Store LogKey name as a value to avoid generating new string instances

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Store LogKey name as a value to avoid generating new string instances
### Why are the changes needed?

To save memory usage on getting the names of `LogKey`s.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 1]
```diff
@@ -16,10 +16,14 @@
...
 package org.apache.spark.internal
 
+import java.util.Locale
+
 /**
  * All structured logging `keys` used in `MDC` must be extends `LogKey`
  */
-trait LogKey
+trait LogKey {
+  val name: String = this.toString.toLowerCase(Locale.ROOT)
```
Comment (gengliangwang) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:25]: We can also make it `lazy val` here. I am not sure if the cost will be eventually higher. I suggest keeping it simple for now.

**Comment by gengliangwang** [comment]
cc @panbingkun

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @gengliangwang .

**Comment by panbingkun** [comment]
Great!
+1, LGTM. Thanks!

**Comment by LuciferYang** [comment]
late LGTM

**Pull request closed without merging**